x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/telemetry/StatsMap.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.telemetry;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * A map to provide tracking incrementing statistics.
+ *
+ * @param <Input> The input to derive the keys and values for the map
+ * @param <SerializedValue> The serializable version of the value stored in the map. This type should be a class for handling things like
+ *                         serializing the value to {@link org.elasticsearch.xcontent.XContent} or within an
+ *                         {@link org.elasticsearch.common.io.stream.StreamInput}
+ * @param <Value> The type of the values stored in the map
+ */
+public class StatsMap<Input, SerializedValue, Value extends Stats & Transformable<SerializedValue>> {
+
+    private final ConcurrentMap<String, Value> stats = new ConcurrentHashMap<>();
+    private final Function<Input, String> keyCreator;
+    private final Function<Input, Value> valueCreator;
+
+    /**
+     * @param keyCreator a function for creating a key in the map based on the input provided
+     * @param valueCreator a function for creating a value in the map based on the input provided
+     */
+    public StatsMap(Function<Input, String> keyCreator, Function<Input, Value> valueCreator) {
+        this.keyCreator = Objects.requireNonNull(keyCreator);
+        this.valueCreator = Objects.requireNonNull(valueCreator);
+    }
+
+    /**
+     * Increment the counter for a particular value in a thread safe manner.
+     * @param input the input to derive the appropriate key in the map
+     */
+    public void increment(Input input) {
```

Comment:
A better way to design this might be to have this method simply be:

```
public void execute(KeyValueFactory factory, Input input, Operation op);
```

My thought with that would be we could use the same map instance to contain stats for various types of metrics. Right now the function for creating the key and value is passed in the map's constructor so we'd probably need completely separate maps to model different metrics.

By using an `execute()` method we could avoid having to add a new _operation_ to this calls for each type of metric we want to do something with. Instead we'd create a class for each operation, and the operation would know how to interact with the `Value` class directly. For example:

```
interface Operation {
  void execute(Stats stats);
}

class IncrementOperation implements Operation {
  public void execute(Stats stats) {
    stats.increment();
  }
}
```

I chose not to go that round initially because it adds more abstraction without knowing if that's actually where this is headed. I'm happy to do it if we're confident it'll help though. Maybe that'll be more clear once we add a few more metrics to track.

Reply:
Thanks for providing this context!